File: 440_KthSmallestInLexicographicalOrder.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int findKthNumber(int n, int k) {
        long curr = 1;
        int newK = k, nextAtSameLevel=1;
        while (newK>1){
            int count = getNumsInSubtree(curr, n);
            if (count<newK){
                newK -= count;
                curr += 1;
            } else {
                curr*=10;
                newK-=1;
            }
        }

        return curr;
    }

private:
    int getNumsInSubtree(long curr, const int& n){
        long nxt = curr+1, count = 0;
        while (curr<=n){
            nxt = min(nxt, n+1l);
            count += nxt-curr;
            curr*=10;
            nxt*=10;
        }

        return count;
    }
};

int main(){
    Solution s;
    cout<<s.findKthNumber(13, 3);
    return 0;
}


/*
Input: n = 13, k = 2
Output: 10
[1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9]

[1]*[-,0..3]
[2]*[-]
[3]*[-]
[4]*[-]
[5]*[-]
[6]*[-]
[7]*[-]
[8]*[-]
[9]*[-]


200
1*[-,0..9]*[-,0..9]

*/

File: 440_KthSmallestInLexicographicalOrder.py
class Solution(object):
    def findKthNumber(self, n, k):
        curr = 1
        k -= 1

        while k > 0:
            step = self.countSteps(n, curr, curr + 1)
            if step <= k:
                curr += 1
                k -= step
            else:
                curr *= 10
                k -= 1

        return curr

    def countSteps(self, n, prefix1, prefix2):
        steps = 0
        while prefix1 <= n:
            steps += min(n + 1, prefix2) - prefix1
            prefix1 *= 10
            prefix2 *= 10
        return steps

# print(Solution().findKthNumber(13,2)) #10
# print(Solution().findKthNumber(2,2)) #2
print(Solution().findKthNumber(13,8)) #4
